test/e2e/multi_rook_ceph_test.go

```diff
@@ -0,0 +1,252 @@
+	"fmt"
+
+	"github.com/cybozu-go/mantle/test/util"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+type multiRookCephTest struct {
+	poolName          string // use the same pool name for both clusters
+	storageClassName1 string
+	storageClassName2 string
+	tenantNamespace1  string
+	tenantNamespace2  string
+
+	pvcName           string // use the same PVC name for both tenants
+	mantleBackupName  string // use the same MantleBackup name for both tenants
+	mantleRestoreName string // use the same MantleRestore name for both tenants
+}
+
+func multiRookCephTestSuite() {
+	test := &multiRookCephTest{
+		poolName:          util.GetUniqueName("pool-"),
+		storageClassName1: util.GetUniqueName("sc-"),
+		storageClassName2: util.GetUniqueName("sc-"),
+		tenantNamespace1:  util.GetUniqueName("ns-"),
+		tenantNamespace2:  util.GetUniqueName("ns-"),
+
+		pvcName:           util.GetUniqueName("pvc-"),
+		mantleBackupName:  util.GetUniqueName("backup-"),
+		mantleRestoreName: util.GetUniqueName("restore-"),
+	}
+
+	Describe("setup environment", test.setupEnv)
+	Describe("test main", test.testMain)
+	Describe("tearDown environment", test.teardownEnv)
+}
+
+func (test *multiRookCephTest) setupEnv() {
+	It("setting up the test environment", func() {
+		fmt.Fprintf(GinkgoWriter, "%+v\n", *test)
+	})
+
+	It("creating common resources", func() {
+		for _, ns := range []string{test.tenantNamespace1, test.tenantNamespace2} {
+			err := createNamespace(ns)
+			Expect(err).NotTo(HaveOccurred())
+		}
+
+		err := applyRBDPoolAndSCTemplate(cephCluster1Namespace, test.poolName, test.storageClassName1)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = applyRBDPoolAndSCTemplate(cephCluster2Namespace, test.poolName, test.storageClassName2)
+		Expect(err).NotTo(HaveOccurred())
+	})
+}
+
+func (test *multiRookCephTest) teardownEnv() {
+	for _, ns := range []string{test.tenantNamespace1, test.tenantNamespace2} {
+		ns := ns
+		It("delete resources in the namespace: "+ns, func() {
+			err := deleteNamespacedResource(ns, "mantlerestore")
+			Expect(err).NotTo(HaveOccurred())
+			err = deleteNamespacedResource(ns, "mantlebackup")
+			Expect(err).NotTo(HaveOccurred())
+			err = deleteNamespacedResource(ns, "pvc")
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("delete the namespace: "+ns, func() {
+			_, stderr, err := kubectl("delete", "namespace", ns)
+			Expect(err).NotTo(HaveOccurred(), string(stderr))
+		})
+	}
+
+	It("clean up the SCs and RBD pools", func() {
+		for _, sc := range []string{test.storageClassName1, test.storageClassName2} {
+			_, stderr, err := kubectl("delete", "sc", sc)
+			Expect(err).NotTo(HaveOccurred(), string(stderr))
+		}
+
+		for _, ns := range []string{cephCluster1Namespace, cephCluster2Namespace} {
+			err := removeAllRBDImageAndSnap(ns, test.poolName)
+			Expect(err).NotTo(HaveOccurred())
+
+			_, _, err = kubectl("delete", "-n", ns, "cephblockpool", test.poolName, "--wait=false")
+			Expect(err).NotTo(HaveOccurred())
```

Comment:
It's better to continue the teardown even if some deletion failed not to restore clean environment as possible.

```suggestion
		It("delete resources in the namespace: "+ns, func() {
			err := deleteNamespacedResource(ns, "mantlerestore")
			err = deleteNamespacedResource(ns, "mantlebackup")
			err = deleteNamespacedResource(ns, "pvc")
		})

		It("delete the namespace: "+ns, func() {
			_, stderr, err := kubectl("delete", "namespace", ns)
		})
	}

	It("clean up the SCs and RBD pools", func() {
		for _, sc := range []string{test.storageClassName1, test.storageClassName2} {
			_, stderr, err := kubectl("delete", "sc", sc)
		}

		for _, ns := range []string{cephCluster1Namespace, cephCluster2Namespace} {
			err := removeAllRBDImageAndSnap(ns, test.poolName)

			_, _, err = kubectl("delete", "-n", ns, "cephblockpool", test.poolName, "--wait=false")
```

Reply:
I would like to leave these because the errors raised by tearDown sometimes made me notice cases that I was not noticed. Of course, if you know it will be an error, then it's not.

Reply:
OK, I understood.